[respeaker_ros] Add publish_multichannel option for publishing multi channel audio

iory opened the pull request:
# What is this?

This PR adds `~publish_multichannel` option for publishing multi-channel audio.

Along with that, there is a change to publish `audio_common_msgs/AudioInfo`.
This change does not affect those who were originally using it on channel 1.

knorth55 commented on respeaker_ros/launch/sample_respeaker.launch:
```diff
@@ -13,23 +13,30 @@
   <node name="speech_to_text" pkg="respeaker_ros" type="speech_to_text.py"
         output="screen">
-    <remap from="audio" to="$(arg audio)"/>
+    <remap from="audio" to="/speech_audio"/>
```
is this change really necessary?

iory replied:
Sorry, I mistakenly commit it.
https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/206aaaadb7e1521fb0cd1db651be5f11f6e7cc3b
I modified it.
I'll rebase if necessary. Or I hope squash merge.

708yamaguchi commented on respeaker_ros/launch/sample_respeaker.launch:
```diff
@@ -13,14 +13,20 @@
-        respawn="true" respawn_delay="10" />
+        respawn="true" respawn_delay="10" >
+    <rosparam subst_value="true" >
+      publish_multichannel: $(arg publish_multichannel)
```
You may forget to add
```
<arg name="publish_multichannel" default="false"/>
```

This type of error cannot be detected by GitHub Action.
This is because in GitHub Action we cannot use respeaker and we do not run `sample_respeaker.launch`

My suggestion is to use `roslaunch_add_file_check()` in `if(CATKIN_ENABLE_TESTING)` section.
```
  find_package(roslaunch REQUIRED)
  roslaunch_add_file_check(launch/sample_respeaker.launch)
```

When I use this check to the current `sample_respeaker.launch`, I get the following error.
```
FAILURE:
[/home/naoya/ros/melodic/src/jsk-ros-pkg/jsk_3rdparty/respeaker_ros/launch/sample_respeaker.launch]:
	[/home/naoya/ros/melodic/src/jsk-ros-pkg/jsk_3rdparty/respeaker_ros/launch/sample_respeaker.launch] requires the 'publish_multichannel' arg to be set
```

c.f. https://github.com/jsk-ros-pkg/jsk_robot/blob/eac815173b8f0e0df3a1f230dd9b68ec937fddfe/jsk_fetch_robot/jsk_fetch_startup/CMakeLists.txt#L78

iory replied:
https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/305343a81a5bc9698443a5c47c32ce8a22b1a3c0
Thanks for your comment.
I added the test code.

708yamaguchi commented on respeaker_ros/launch/sample_respeaker.launch:
```diff
@@ -13,23 +13,30 @@
   <!-- self cancellation -->
   <arg name="self_cancellation" default="true"/>
+  <!-- audio info topic name -->
+  <arg name="audio_info" default="audio_info"/>
```
Setting `audio_info` to `"audio_info"` may change the default behavior.
In fact, `sample_respeaker.test` fails with this PR.[

Setting `audio_info` to `"false"` is correct?
```
<arg name="audio_info" default="false"/>
```

iory replied:
Thanks for your comment!
I modified the test code's argument.
Ideally, the default is to output audio_info and the user does not need to give parameters such as samplerate.

708yamaguchi replied:
OK, I see.

I think it is good to publishing audio_info by default

708yamaguchi commented on respeaker_ros/scripts/speech_to_text.py:
```diff
@@ -2,24 +2,46 @@
         # format of input audio data
-        self.sample_rate = rospy.get_param("~sample_rate", 16000)
-        self.sample_width = rospy.get_param("~sample_width", 2)
+        if rospy.get_param('~audio_info', None):
```
```
if rospy.get_param('~audio_info', false):
```
is more readable?

iory replied:
https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/296cc6ae3c326ec4cdf6c8f8487155017fb56ffb
I modified the code to make it more readable.
I look forward to your feedback.

708yamaguchi replied:
In this case, `'~audio_info'` rosparam only takes string variable.
I think it's good.

iory replied:
Thank you!

708yamaguchi commented:
I think it is desirable to use this feature on real robot (e.g. Fetch) before merging.

tkmtnt7000 reviewed (commented):
I think so, too. We have to confirm that this node works well on real robots before merging.
Anyway, that's a wonderful node. I hope this node to solve the problem around voice recognition. e.g: We have to speak voice commands extremely near the robot.

tkmtnt7000 commented on respeaker_ros/scripts/respeaker_node.py:
```diff
@@ -322,21 +328,38 @@ def __init__(self):
-        self.respeaker_audio = RespeakerAudio(self.on_audio, suppress_error=suppress_pyaudio_error)
+        self.respeaker_audio = RespeakerAudio(self.on_audio, suppress_error=suppress_pyaudio_error,
+                                              publish_multichannel=self.publish_multichannel)
+        self.n_channel = 1
+        if self.publish_multichannel:
+            self.n_channel = self.respeaker_audio.channels
```
I am sorry for additional reviewing.
I think it would be nice to write a brief background at some positions as to why `~publish_multichannel` parameter was added (or why this node starts to support multi channel) for future developers.

tkmtnt7000 replied:
Thank you!
It's very clear and easy to understand for future developers.

iory commented:
Removed `~publish_multichannel` option and I modified the node to make publish `~audio_raw` and `~speech_audio_raw` for publishing multi-channel audio.

This change does not affect those who were originally using it on channel 1 because `~audio` output does not change.

708yamaguchi commented:
Thanks.

This PR has many good features, so I would appreciate it
- to add README about pub/sub topics and rosparams. (I am very sorry that I did not add these information before...)
- to split the pull request. (Not necessary, but it is a bit difficult for me to check big diffs...)
For example,
  - PR1: add audio info 
  - PR2: add multichannel
  - PR3: add merged playback

iory commented:
@708yamaguchi 
Thanks for your comments.

> to add README about pub/sub topics and rosparams. (I am very sorry that I did not add these information before...)

I added the commits to reflect your comments.

> to split the pull request. (Not necessary, but it is a bit difficult for me to check big diffs...)
> For example,
> PR1: add audio info
> PR2: add multichannel
> PR3: add merged playback

This PR is for multichannel output.
When using multichannel, user does not know the number of channels from the information only for audio. Therefore audio_info is required. (e.g. PR2 depends on PR1).
As the merged playback information has already been obtained when outputting the multichannel, it's clear to publish merged playback in this PR.

For you, I wrote the document, https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/92477a52e0ece50853271d1596409b2ebbb8bb6a https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/8caae81693ea8c7bb6d41aec397aaa21e5add2b3 https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/383670f44610c6524dccde2959da7d04e3c69d42
　and wrote the comment in codes.
I'm looking forward to your feedback.

708yamaguchi commented on respeaker_ros/scripts/respeaker_node.py:
```diff
@@ -322,21 +321,80 @@ def __init__(self):
+                "speech_audio_raw", AudioData, queue_size=10)
+            self.speech_raw_prefetch_bytes = int(
+                self.n_channel - 2
+                * self.speech_prefetch
```
Is `(self.n_channel - 2)` correct?

In my ipython, I got the following result.
```
In [1]: int(
   ...: 6 - 2
   ...: * 5)
Out[1]: -4
```

iory replied:
Thanks!
I fixed it.
https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/4b99d012c196e155c4f97b4180d449cd4bacda5c

708yamaguchi commented on respeaker_ros/scripts/respeaker_node.py:
```diff
@@ -16,6 +16,7 @@
 import sys
 import time
 from audio_common_msgs.msg import AudioData
+from audio_common_msgs.msg import AudioInfo
```
`AudioInfo` is not released on indigo, maybe.
https://github.com/ros-gbp/audio_common-release

iory replied:
I see.
`jsk_3rdparty`'s test drop the indigo test.
I added the code to avoid AudioInfo import error.
https://github.com/jsk-ros-pkg/jsk_3rdparty/pull/350/commits/2f87dbbabb22f7d1975bd0d603ff059b2784c77f

708yamaguchi commented:
@iory 

Thank you for your quick update!
I check this code as much as I can.

In addition, please check the behavior of `respeaker_node.py` with the following branches.
For fetch, https://github.com/knorth55/jsk_3rdparty/pull/23
For PR2, https://github.com/knorth55/jsk_3rdparty/pull/24

iory commented:
Thank you.
I'm testing on pr1040 and fetch15.
https://github.com/708yamaguchi/jsk_3rdparty/pull/4
https://github.com/708yamaguchi/jsk_3rdparty/pull/3

708yamaguchi commented:
I am very sorry, but I had tested this pull request in fetch, but not in pr2.
This is because I left the following pull request unmerged for a long time.
https://github.com/708yamaguchi/jsk_3rdparty/pull/4

At least this PR is working well with fetch, so I will look at PR2 for a few days and if it looks OK, I will approve it.

708yamaguchi reviewed (approved):
PR2 has not output any respeaker-related errors for the past three days.
I check that `/audio` and `/speech_to_text` seems good.

iory commented:
Thanks!

k-okada reviewed (commented):
please proved linke to demo that requres multichannel audio
